torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+from torch.fx import Graph, Node, symbolic_trace
+
+import collections
+from typing import Callable, Dict, List, Set
+
+# `anchor` - node from which the match was found
+# `nodes_map` - maps nodes in the pattern subgraph to nodes in the
+#               larger graph
+Match = collections.namedtuple("Match", "anchor nodes_map")
+
+class SubgraphMatcher:
+    def __init__(self, pattern: Graph) -> None:
+        self.pattern = pattern
+        try:
+            self.pattern_anchor = next(iter(reversed(pattern.nodes)))
+        except Exception:
+            raise ValueError("SubgraphMatcher cannot be initialized with an "
+                             "empty pattern")
+        # Maps nodes in the pattern subgraph to nodes in the larger graph
+        self.nodes_map: Dict[Node, Node] = {}
+
+    def matches_subgraph_from_anchor(self, anchor: Node) -> bool:
+        """
+        Check if the whole pattern can be matched starting from `anchor` in
+        the larger graph
+        """
+        self.nodes_map = {}
+        if (self.match_nodes(self.pattern_anchor, anchor)):
+            return True
+        return False
+
+    def match_nodes(self, n1: Node, n2: Node) -> bool:
+        """
+        Compare Nodes `n1` (pattern node) and `n2` (graph node) for equality
+        """
+        # Check if we've already matched these nodes in the current
+        # traversal
+        if n1 in self.nodes_map:
+            return self.nodes_map[n1] == n2
+
+        # Terminate early if the node attributes are not equal
+        if ((n1.op != n2.op and n1.op != "output")
+            or (n1.op == "call_function" and n1.target != n2.target)
+            or (n1.op == "placeholder" and n2.op == "placeholder"
+                and n1.name != n2.name)):
+            return False
+
+        # Optimistically mark `n1` as a match for `n2`
+        self.nodes_map[n1] = n2
+
+        # Traverse the use-def relationships to ensure that `n1` is a true
+        # match for `n2`
+        for n1_ in n1.uses.keys():
+            match_found = False
+            for n2_ in n2.uses.keys():
+                if match_found or self.match_nodes(n1_, n2_):
+                    match_found = True
+            if not match_found:
+                self.nodes_map.pop(n1)
+                return False
+
+        return True
+
+
+def replace_pattern(original_graph: Graph, pattern : Callable, replacement : Callable) -> None:
+    """
+    Function that takes one set of operators and their data dependencies
+    and replaces them with another
```

Comment:
Since this is the top-level API, do you mind:

* Formatting this (with Args etc) according to the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)
* Adding a quick example to the docstring